lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,517 @@
+# -*- coding: utf-8 -*-
+# Copyright European Organization for Nuclear Research (CERN) since 2012
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Generic decommissioning profile."""
+
+import logging
+from datetime import datetime, timedelta
+from enum import Enum
+from typing import TYPE_CHECKING
+
+from sqlalchemy.exc import NoResultFound
+
+from rucio.common.config import config_get_int
+from rucio.common.exception import (Duplicate, RequestNotFound, RuleNotFound,
+                                    RuleReplaceFailed, ReplicaNotFound,
+                                    UnsupportedOperation)
+from rucio.core.lock import get_replica_locks, get_replica_locks_for_rule_id
+from rucio.core.replica import (list_replicas_per_rse, set_tombstone,
+                                update_replica_state)
+from rucio.core.request import get_request_by_did
+from rucio.core.rse import add_rse_attribute, update_rse
+from rucio.core.rule import (list_rules_for_rse_decommissioning, move_rule,
+                             update_rule)
+from rucio.db.sqla.constants import ReplicaState
+
+from ..config import DecommissioningStatus, set_status
+from ..throttling import DeletionLimitExceeded, LimitChecker
+
+if TYPE_CHECKING:
+    from collections.abc import Callable, Iterable
+    from typing import Any
+    from uuid import UUID
+
+    Initialization = Callable[[dict[str, Any], DecommissioningStatus], None]
+    RulesListing = Callable[[UUID], Iterable[dict[str, Any]]]
+    RuleSorting = Callable[[dict[str, Any], dict[str, Any]], 'ActionDecision']
+    RuleAction = Callable[[dict[str, Any], dict[str, Any], 'ActionDecision',
+                           LimitChecker], None]
+    Finalization = Callable[[dict[str, Any]], None]
+
+
+class ActionDecision(Enum):
+    """Action (or no action) flags for rules."""
+    DONT_REMOVE = 0
+    CAN_REMOVE = 1
+    BEING_REMOVED = 2
+
+
+def generic_profile_template_function(
+    rse: 'dict[str, Any]',
+    config: 'dict[str, Any]',
+    initialization: 'Initialization',
+    rules_listing: 'RulesListing',
+    rule_sorting: 'RuleSorting',
+    rule_action: 'RuleAction',
+    finalization: 'Finalization',
```

Comment:
Changed the argument names to initializer, finalizer, and discoverer (previously rule_sorting).

As for factoring decision and action out, I think there is still a benefit of code reusability. With the previous structure, there were a lot of duplicate code between generic_delete and generic_move. From what you told me about ATLAS, I would say that what we need is an even finer factoring. The new commit includes a proposal in this direction.

The new profile function is passed a list of (condition function, action function) tuples. In this structure, the conditions are evaluated against each rule in the order given in the list, and the action paired with the first matching condition is taken. The previous `rule_sorting` function was anyway evaluating different conditions one by one and returning whenever something matched, so there is no change in the flow. The new structure however allows the ATLAS profile to insert a condition-action pair that looks for the "log" data type and declares the files as bad.

Please let me know if you like the new structure; if we decide to go this way, I'll add annotations etc. and finalize the daemon. Also, if we decide to use the new scheme, LimitChecker and throttling mechanism can be simplified because there is no more need to raise an exception from deep within some function to stop the decommissioning processing. Throttling can be done from within the top-level `generic_profile_template_function` using simple counters.

Reply:
I find that your approach has a lot of merit and I approve of it. Please proceed.

Reply:
I went ahead with the proposed approach. In addition, I figured that the "generic profile template function" is generic enough that whatever profiles that we might define in the future should be using it, so I moved the function to higher level (i.e. outside the `profiles`) directory. The "profile" functions such as `generic_delete` now returns a dict specifying the initializer, discoverer, handlers, and finalizer. The top-level daemon function gets these functions from the profile and calls the generic template function with them.